src/Elastic.Apm/Metrics/IMetricsProvider.cs

```diff
@@ -0,0 +1,32 @@
+using System.Collections.Generic;
+using Elastic.Apm.Api;
+
+namespace Elastic.Apm.Metrics
+{
+	/// <summary>
+	/// Defines an interface that every class implements that provides some metric value.
+	/// This interface is known to the <see cref="MetricsCollector" /> type and you
+	/// can implement new providers for other metrics by implementing this interface
+	/// and adding it to <see cref="MetricsCollector" />.
+	/// </summary>
+	internal interface IMetricsProvider
+	{
+		/// <summary>
+		/// Stores the number of calls to the <see cref="GetValue"/> method when it either returned null or an empty list.
+		/// This is used by <see cref="MetricsCollector"/>
+		/// </summary>
+		int ConsecutiveNumberOfFailedReads { get; set; }
+
+		/// <summary>
+		/// The name that refers to the provider in the logs. E.g. "total process CPU time".
+		/// Make sure this is human understandable and tells the reader what type of value this provider is intended to provider.
+		/// </summary>
+		string NameInLogs { get; }
+
+		/// <summary>
+		/// The main part of the provider, the implementor should do the work to read the value(s) of the given metric(s) in this method.
+		/// </summary>
+		/// <returns>The key and the value of the metric(s)</returns>
+		IEnumerable<Sample> GetValue();
```

Comment:
I don’t know.. we should ask the server developers.

Reply:
[APM server spec](https://github.com/elastic/apm-server/blob/master/docs/spec/metricsets/metricset.json) doesn't allow it because `samples` is a map (JSON object). I am okay with .NET agent spec saying if more than sample uses the same metric name the agent will choose one (without specifying which one exactly) - I just want to make sure we don't crash in that case which I'm afraid is what might happen because `MetricSetConverter.WriteJson` [calls](https://github.com/elastic/apm-agent-dotnet/pull/248/files/736e2f70664f590e36a9c75aa24950e672d7722e#diff-2f91ae0b37450385a003a95280872758R19) `samples.Add(item.KeyValue.Key, valueObj);` and `Dictionary` throws `ArgumentException` if element with the same key already exists. The solution might be to check if `samples` already contains that key and skip it if it does.

Reply:
Added a check to make sure we only use each key once - with this the 1. one will win, the rest is ignored. 

>  I just want to make sure we don't crash in that case which I'm afraid is what might happen because

Just for the record: it would not have been a crash, the serialization logic is try-catch-ed.